core/logger/audit/audit_logger.go

```diff
@@ -0,0 +1,312 @@
+			l.logger.Errorw("Could not send log to audit subsystem because it has gone away!")
+		} else {
+			l.logger.Errorw("Audit log buffer is full. Dropping log with eventID: %s", eventID)
+		}
+	}
+}
+
+// Start the audit logger and begin processing logs on the channel
+func (l *AuditLoggerService) Start(context.Context) error {
+	if !l.enabled {
+		return errors.Errorf("The audit logger is not enabled")
+	}
+	go l.runLoop()
+	return nil
+}
+
+// Stops the logger and will close the channel.
+func (l *AuditLoggerService) Close() error {
+	if !l.enabled {
+		return errors.Errorf("The audit logger is not enabled")
+	}
+
+	l.logger.Warnf("Disabled the audit logger service")
+	l.cancel()
+	<-l.chDone
+
+	return nil
+}
+
+func (l *AuditLoggerService) Healthy() error {
+	if !l.enabled {
+		return errors.Errorf("The audit logger is not enabled")
+	}
+
+	if len(l.loggingChannel) == bufferCapacity {
+		return errors.Errorf("The audit log buffer is full")
+	}
+
+	return nil
+}
+
+func (l *AuditLoggerService) Ready() error {
+	if !l.enabled {
+		return errors.Errorf("The audit logger is not enabled")
+	}
+
+	return nil
+}
+
+// Entrypoint for our log handling goroutine. This waits on the channel and sends out
+// logs as they come in.
+//
+// This function calls postLogToLogService which blocks.
+func (l *AuditLoggerService) runLoop() {
+	defer close(l.chDone)
+
+	for {
+		select {
+		case <-l.ctx.Done():
+			// I've made this an error since we expect it should never happen.
+			l.logger.Errorf("The audit logger has been requested to shut down!")
+			return
+		case event := <-l.loggingChannel:
+			l.postLogToLogService(event.eventID, event.data)
+		}
+	}
+}
+
+// Takes an EventID and associated data and sends it to the configured logging
+// endpoint. This function blocks on the send by timesout after a period of
+// several seconds. This helps us prevent getting stuck on a single log
+// due to transient network errors.
+//
+// This function blocks when called.
+func (l *AuditLoggerService) postLogToLogService(eventID EventID, data Data) {
+	// Audit log JSON data
+	logItem := map[string]interface{}{
+		"eventID":  eventID,
+		"hostname": l.hostname,
+		"localIP":  l.localIP,
+		"env":      l.environmentName,
+		"data":     data,
+	}
+
+	// Optionally wrap audit log data into JSON object to help dynamically structure for an HTTP log service call
+	if l.jsonWrapperKey != "" {
+		logItem = map[string]interface{}{l.jsonWrapperKey: logItem}
+	}
+
+	serializedLog, err := json.Marshal(logItem)
+	if err != nil {
+		l.logger.Errorw("Unable to serialize wrapped audit log item to JSON", "err", err, "logItem", logItem)
+		return
+	}
+
+	// Send to remote service
+	req, err := http.NewRequest("POST", l.forwardToUrl, bytes.NewReader(serializedLog))
+	if err != nil {
+		l.logger.Errorf("Failed to create request to remote logging service!")
+	}
+	for _, header := range l.headers {
+		req.Header.Add(header.Header, header.Value)
+	}
+	resp, err := l.loggingClient.Do(req)
```

Comment:
This should take the context, so that any in-flight requests are cancelled immediately on application shutdown.

Reply:
I think it will need to happen on 268 via https://pkg.go.dev/net/http#NewRequestWithContext

Reply:
Done